---
output:
  pdf_document: default
  html_document: default
---

% !TeX program = pdfLaTeX
---
title: "Analyses ofE14 Electroporated Cells with Seurat Workflow"
author: "Bismark Appiah"
date: "7 April 2020"
output:
  html_document:
    df_print: paged
---


#Loading data matrix  
Convert data to "SingleCellExperiment" Class.

```{r loading, echo=TRUE} 
data2 <- readRDS("EUE14ALL.RData")
sc2 <- SingleCellExperiment(list(counts=as.matrix(data2)))
sc2
EUE14DMS_1 <- sc2[, grep('EUEE14DMS', colnames(sc2))]
EUE14DMS <- colnames(EUE14DMS_1)

EUE14EPZ_1 <- sc2[, grep('EUEE14EPZ', colnames(sc2))]
EUE14EPZ <- colnames(EUE14EPZ_1)
```


## Define DMSO and EPZ conditions  

```{r metadata, include=TRUE, echo=TRUE}

sc2$condition <- "NA"
sc2$condition[which(colnames(sc2) %in% EUE14DMS)] <- "DMSO"
sc2$condition[which(colnames(sc2) %in% EUE14EPZ)] <- "EPZ"
table(sc2$condition)
```

## Add mitochondrial genes and ERCC spike-ins to sc2 object  

There were no ERCC spike-ins used in the experiment.
```{r spikeins, include=TRUE, echo=TRUE}
isSpike(sc2, "MT") <- rownames(sc2)[grep("^(mt)",rownames(sc2),invert=FALSE)]
isSpike(sc2, "ERCC") <- rownames(sc2)[grep("^(ERCC)",rownames(sc2),invert=FALSE)]
sc2
```

# Quality control  

## Remove genes that are not expressed  

```{r removegenes, include=TRUE, echo=TRUE}
keep_feature <- rowSums(counts(sc2) > 0) > 0
sc2 <- sc2[keep_feature,]
```

#Calculate quality metrics  

There were no ERCC spike-ins used.

```{r qualitymetrics, include=TRUE, echo=TRUE}
sc2 <- calculateQCMetrics(sc2)
```

```{r qualplot, include=TRUE, fig.height=10, fig.width=10}
multiplot(cols=2,
    plotColData(sc2, x="condition", y="total_counts"),
    plotColData(sc2, x="condition", y="total_features_by_counts"),
    plotColData(sc2, x="condition", y="pct_counts_MT")
)
```
Total counts and distribution of counts for cells in both looks quite good
We can proceed with further analysis



## Visualize library sizes  

```{r, include=TRUE}
hist(
    sc2$total_counts,
    breaks = 100
)
abline(v = 2500, col = "red")
```

## Filter cells by library size > 2500  

```{r, include=TRUE}
filter_by_total_counts <- (sc2$total_counts > 2500)
table(filter_by_total_counts)
```

```{r, include=TRUE}
sc2 <- sc2[,filter_by_total_counts]
```

## Visualize the total number of unique genes detected in each sample  

```{r, include=TRUE}
hist(
    sc2$total_features_by_counts,
    breaks = 100
)
abline(v = 1000, col = "red")
```

## Filter cells by expressed features > 1000  

```{r, include=TRUE}
filter_by_expr_features <- (sc2$total_features_by_counts > 1000)
table(filter_by_expr_features)
```

```{r, include=TRUE}
sc2 <- sc2[,filter_by_expr_features]
```

Cells with outlier values for mitochondrial genes are identified based on some number of MADs from the median value.

```{r, include=TRUE}
high.mt <- isOutlier(sc2$pct_counts_MT, nmads=3, type="higher", batch=sc2$condition)
data.frame(HighMito=sum(high.mt))
```

We only retain cells that pass all of the specified criteria.
Of course, this involves some assumptions about independence from biology. 
(For example, don't use the mitochondrial proportions if the number/activity of mitochondria changes between cell types.)

```{r, include=TRUE}
discard <- high.mt
data.frame(TotalLost=sum(discard), TotalLeft=sum(!discard))
```

We toss out the cells that we consider to be low-quality, and keep the rest.

```{r, include=TRUE}
sc2 <- sc2[,!discard]
ncol(sc2)
```

https://scrnaseq-course.cog.sanger.ac.uk/website/cleaning-the-expression-matrix.html#exprs-qc

```{r, include=TRUE}
plotColData(sc2, x = "total_features_by_counts", y = "pct_counts_MT")
```

# Examining the genes

We inspect the distribution of log-mean counts across all genes.
The peak represents the bulk of moderately expressed genes while the rectangular component corresponds to lowly expressed genes.

```{r, include=TRUE}
ave.counts <- calcAverage(sc2)
hist(log10(ave.counts), breaks=100, main="", col="grey80",
    xlab=expression(Log[10]~"average count"))
```

We also look at the identities of the most highly expressed genes.
This should generally be dominated by constitutively expressed transcripts, such as those for ribosomal or mitochondrial proteins.
The presence of other classes of features may be cause for concern if they are not consistent with expected biology.
#Show top 30 of most highly expressed genes
```{r}
plotHighestExprs(sc2, n=30)
```

```{r, include=TRUE}
clusters <- quickCluster(sc2,use.ranks=FALSE)
sc2 <- computeSumFactors(sc2, cluster=clusters)
```

From: Lun et al. (2016) A step-by-step workflow for low-level analysis of single-cell RNA-seq data with Bioconductor.

"In this case, the size factors are tightly correlated with the library sizes for all cells. This suggests that the
systematic differences between cells are primarily driven by differences in capture efficiency or sequencing depth. Any
DE between cells would yield a non-linear trend between the total count and size factor, and/or increased scatter around
the trend. This does not occur here as strong DE is unlikely to exist within a homogeneous population of cells."

In our case there is also a linear trend, but with some scatter around it.

```{r, include=TRUE}
plot(sizeFactors(sc2), sc2$total_counts/1e3, log="xy",
ylab="Library size (thousands)", xlab="Size factor")
```

```{r, include=TRUE}
sc2 <- scater::normalize(sc2)
sc2
```


```{r, include=TRUE}
var.fit <- trendVar(sc2, method="loess", loess.args=list(span=0.05), use.spikes=FALSE)
var.out <- decomposeVar(sc2, var.fit)
head(var.out)    
```

"We can have a look at the fitted trend.
Some tinkering may be required to get a good fit, usually by modifying `span=`."

```{r hvgplothsc, include=TRUE}
plot(var.out$mean, var.out$total, pch=16, cex=0.6, xlab="Mean log-expression", 
    ylab="Variance of log-expression")
curve(var.fit$trend(x), add=TRUE, col="dodgerblue", lwd=2)
cur.spike <- isSpike(sc2)
points(var.out$mean[cur.spike], var.out$total[cur.spike], col="red", pch=16)
```

```{r, include=TRUE}
#hvg.out <- var.out[which(var.out$FDR <= 0.05 & var.out$bio >= 0.25),]
#hvg.out <- var.out[which(var.out$FDR <= 0.05 & var.out$bio >= 0.01),]
hvg.out <- var.out[with(var.out, order(bio,FDR, decreasing = TRUE)),][1:800,]
hvg.out <- hvg.out[order(hvg.out$bio, decreasing = TRUE),]
nrow(hvg.out)
```


# Checking the distribution of expression values for the top HVGs to ensure that the variance estimate is not being dominated by one or two outlier cells.
```{r, include=TRUE}
plotExpression(sc2, rownames(hvg.out)[1:10])
```

## Extract highly variable genes (HVG)  
 
```{r, include=TRUE}
sce_hvg <- sc2[row.names(hvg.out),]
```

# Seurat clustering analysis  

Convert SingleCellExperiment object into Seurat object.

```{r}
rownames(sce_hvg) <- str_remove(rownames(sce_hvg), "[|]")
```


```{r}
sce_hvg.seurat <- as.Seurat(x = sce_hvg)
```

Scaling data before running PCA.

```{r}
all.genes <- rownames(sce_hvg.seurat)
sce_hvg.seurat <- ScaleData(sce_hvg.seurat, features = all.genes)
```

Remove mitochondrial genes.

```{r removeercc}
sce_hvg.seurat <- sce_hvg.seurat[grep("^(mt)",row.names(sce_hvg.seurat),invert=TRUE),]
```


Run PCA on Seurat object.

```{r}
sce_hvg.seurat <- RunPCA(sce_hvg.seurat, npcs = 60, features = rownames(sce_hvg.seurat))
```

# Inspecting principal components
```{r}
VizDimLoadings(sce_hvg.seurat, dims = 1:2, reduction = "pca")
```

```{r}
DimPlot(sce_hvg.seurat, reduction = "pca")
```

```{r}
DimHeatmap(sce_hvg.seurat, dims = 1:10, cells = 500, balanced = TRUE)
```

Heatmap allows to see some structure in the data. It suggest that the first 10 PCs include some structure. These can then be used for the cluster analysis.

# Determine the dimensionality of the dataset  

This part is based on: https://satijalab.org/seurat/v3.0/pbmc3k_tutorial.html

"To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metafeature’ that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many componenets should we choose to include? 10? 20? 100?

In Macosko et al, we implemented a resampling test inspired by the JackStraw procedure. We randomly permute a subset of the data (1% by default) and rerun PCA, constructing a ‘null distribution’ of feature scores, and repeat this procedure. We identify ‘significant’ PCs as those who have a strong enrichment of low p-value features."

```{r}
sce_hvg.seurat <- JackStraw(sce_hvg.seurat, num.replicate = 100, dims = 20)
sce_hvg.seurat <- ScoreJackStraw(sce_hvg.seurat, dims = 1:20)
```

```{r}
JackStrawPlot(sce_hvg.seurat, dims = 1:20, xmax = 0.75, ymax = 1.0)
```



```{r}
ElbowPlot(sce_hvg.seurat, ndims = 15, reduction = "pca")
```

The more approximate elbow plot also suggests to use 8 PCs.

# Cluster evaluation  

Run Seurat clustering with varying resolution parameters to assess cluster stability. Parameter is varied between 0.4 to 1.2 in steps of 0.1. The resulting clusterings are then compared to each other using the adjusted rand index (ARI).

```{r, message=FALSE}
resolution <- seq(0.4, 1.2, 0.1)
sce_hvg.seurat <- FindNeighbors(sce_hvg.seurat, dims = 1:10)
cluster_out <- mclapply(1:length(resolution), function(x){
  sce_hvg.seurat <- FindClusters(sce_hvg.seurat, resolution = resolution[x], algorithm = 1)
  Idents(sce_hvg.seurat)
})

combinations <- combn(1:length(resolution),2)
cluster_ari <- vector(length = ncol(combinations))
for(i in 1:ncol(combinations)){
    cluster_ari[i] <- adjustedRandIndex(cluster_out[[combinations[1,i]]], cluster_out[[combinations[2,i]]]) %>%
      as.data.frame()

    names(cluster_ari[[i]]) <- str_c(as.character(combinations[,i])[1],as.character(combinations[,i])[2], sep="_")
}
cluster_ari <- unlist(cluster_ari)
cluster_ari
```

Compute average silhouette width for resolution parameter.

```{r}
avg_sil_width <- lapply(1:length(cluster_out), function(x){
    sil_out <- cluster::silhouette(as.integer(cluster_out[[x]]),dist(t(log1p(GetAssayData(sce_hvg.seurat)))))
    summary(sil_out[,3])
})
avg_sil_width
```


# Cluster the cells  

Using a resolution of 0.4 seems to be the most stable version according to the average silhouette and ARI.

```{r}
sce_hvg.seurat <- FindNeighbors(sce_hvg.seurat, dims = 1:10)
sce_hvg.seurat <- FindClusters(sce_hvg.seurat, resolution = 0.4, algorithm = 1)
```

```{r}
head(Idents(sce_hvg.seurat), 5)
```

# Run t-SNE for visualization in two dimensions  

```{r}
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages = 'umap-learn')
#sce_hvg.seurat <- RunUMAP(pbmc, dims = 1:10)
```


```{r}
set.seed(254)
sce_hvg.seurat <- RunUMAP(sce_hvg.seurat, dims = 1:8)
DimPlot(sce_hvg.seurat, reduction = "umap")
```

# Finding differentially expressed features (cluster biomarkers)  

## Markers for Cluster 0  

```{r}
# find all markers of cluster 0
cluster0.markers <- FindMarkers(sce_hvg.seurat, ident.1 = 0)
head(cluster0.markers, n = 10)
```

# umap plots of top 10 Cluster 0 markers
```{r fig.height=10, fig.width=10}
FeaturePlot(sce_hvg.seurat, features = c(row.names(cluster0.markers)[1:10]))
```

##Markers for Cluster 1  

```{r}
# find all markers of cluster 1
cluster1.markers <- FindMarkers(sce_hvg.seurat, ident.1 = 1)
head(cluster1.markers, n = 10)
```

#umap plots of top 10 Cluster 1 markers
```{r}
FeaturePlot(sce_hvg.seurat, features = c(row.names(cluster1.markers)[1:10]))
```

##Markers for Cluster 2  

```{r}
# find all markers of cluster 2
cluster2.markers <- FindMarkers(sce_hvg.seurat, ident.1 = 2)
head(cluster2.markers, n = 10)
```

# umap plots of top 10 Cluster 2 markers
```{r fig.height=10, fig.width=15}
FeaturePlot(sce_hvg.seurat, features = c(row.names(cluster2.markers)[1:10]))
```


##Markers for Cluster 3  

```{r}
# find all markers of cluster 3
cluster3.markers <- FindMarkers(sce_hvg.seurat, ident.1 = 3)
head(cluster3.markers, n = 10)
```

# umap plots of top 10 Cluster 3 markers
```{r fig.height=10, fig.width=15}
FeaturePlot(sce_hvg.seurat, features = c(row.names(cluster3.markers)[1:10]))
```

##Markers for Cluster 4  

```{r}
# find all markers of cluster 4
cluster4.markers <- FindMarkers(sce_hvg.seurat, ident.1 = 4)
head(cluster4.markers, n = 10)
```

# umap plots of top 10 Cluster 4 markers
```{r}
FeaturePlot(sce_hvg.seurat, features = c(row.names(cluster4.markers)[1:10]))
```


##Markers for Cluster 5  

```{r}
# find all markers of cluster 5
cluster5.markers <- FindMarkers(sce_hvg.seurat, ident.1 = 5)
head(cluster5.markers, n = 10)
```
# umap plots of top 10 Cluster 5 markers
```{r}
FeaturePlot(sce_hvg.seurat, features = c(row.names(cluster5.markers)[1:10]))
```



#Computing differential markers for each cluster compared to all other cells. Idea is to look at markers highly specific for each cluster. You will find cluster labels in column 6 with corresponding cluster-specific marker gene in column 7 of table below
```{r}
# find markers for every cluster compared to all remaining cells, report only the positive ones
sce_hvg.seurat.markers <- FindAllMarkers(sce_hvg.seurat, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
sce_hvg.seurat.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
```

# Showing top 10 marker genes in each cluster as a heatmap.
```{r fig.height=10, fig.width=10}
top10 <- sce_hvg.seurat.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(sce_hvg.seurat, features = top10$gene) + NoLegend()
```


#umap representation of cells in each condition (DMSO/EPZ)
From here, we can compute proportions of cells for each cluster. This will be included in my next analyses report
```{r fig.height=5, fig.width=10}
DimPlot(sce_hvg.seurat, reduction = "umap", split.by = "condition", pt.size = 1.5)
```



# Using Slingshot for pseudotime analysis  

From: https://www.bioconductor.org/packages/release/bioc/vignettes/slingshot/inst/doc/vignette.html#using-slingshot

```{r}
sce_hvg.slingshot <- as.SingleCellExperiment(sce_hvg.seurat)
```

## Gaussian mixture model for clustering  

```{r}
cl1 <- Mclust(reducedDim(sce_hvg.slingshot)[,1:2])$classification
cl1 <- Idents(sce_hvg.seurat)
colData(sce_hvg.slingshot)$GMM <- cl1

plot(reducedDim(sce_hvg.slingshot)[,1:2], col = brewer.pal(9,"Set1")[cl1], pch=16, asp = 1)
```



```{r}
sce_hvg.slingshot <- slingshot::slingshot(sce_hvg.slingshot, clusterLabels = 'GMM', reducedDim = 'UMAP')
```

```{r}
saveRDS(file="sce_hvg.slingshot", sce_hvg.slingshot)
```

```{r}
sce_hvg.slingshot <- readRDS(file="sce_hvg.slingshot")
```

```{r}
summary(sce_hvg.slingshot$slingPseudotime_1)
```

```{r fig.height=10, fig.width=10}
colors <- colorRampPalette(brewer.pal(11,'Spectral')[-6])(100)
plotcol <- colors[cut(sce_hvg.slingshot$slingPseudotime_1, breaks=100)]

plot(reducedDims(sce_hvg.slingshot)$UMAP, col = plotcol, pch=16, asp = 1)
lines(SlingshotDataSet(sce_hvg.slingshot), lwd=2, col='black')
```


Cell lineages originate from the red cells. So far, the clusters are colored automatically by the slingshot package
I am working on changing this so the colors for the clusters from all preceding analyses match that of the slingshot analyses.
```{r fig.height=10, fig.width=10}
plot(reducedDims(sce_hvg.slingshot)$UMAP, col = brewer.pal(9,'Set1')[sce_hvg.slingshot$seurat_clusters], pch=16, asp = 1)
lines(SlingshotDataSet(sce_hvg.slingshot), lwd=2, type = 'lineages', col = 'black')
```


# Session info  

```{r, echo=TRUE}
sessionInfo()
```







